internal/audit/cerboshub/cerboshub.go

```diff
@@ -171,27 +181,52 @@ func schedule(db *badgerv4.DB, muTimer *mutexTimer, syncer IngestSyncer, minFlus
 	// This prevents a retry completion occurring immediately before the next sync
 	// (and therefore burdening the backend).
 	muTimer.set(minFlushInterval)
+
+	select {
+	case errCh <- err:
+	default:
+	}
 }
 
 func streamLogs(db *badgerv4.DB, syncer IngestSyncer, maxBatchSize, numGo int, flushTimeout time.Duration) error {
+	// We use two streams: one for access logs, and one for decision logs, as this allows us to
+	// avoid the penalty of per-key string inspection when inferring the type down the line.
+	ctx := context.Background()
+
+	p := pool.New().WithContext(ctx).WithCancelOnError().WithFirstError()
+	p.Go(func(ctx context.Context) error {
+		return streamPrefix(ctx, db, syncer, maxBatchSize, numGo, flushTimeout, logsv1.IngestBatch_ENTRY_KIND_ACCESS_LOG)
+	})
+	p.Go(func(ctx context.Context) error {
+		return streamPrefix(ctx, db, syncer, maxBatchSize, numGo, flushTimeout, logsv1.IngestBatch_ENTRY_KIND_DECISION_LOG)
+	})
+
+	return p.Wait()
+}
+
+func streamPrefix(ctx context.Context, db *badgerv4.DB, syncer IngestSyncer, maxBatchSize, numGo int, flushTimeout time.Duration, kind logsv1.IngestBatch_EntryKind) error {
 	// BadgerDB transactions work with snapshot isolation so we only take a view of the DB.
 	// Subsequent writes aren't blocked.
 	stream := db.NewStream()
 	stream.NumGo = numGo
-	stream.Prefix = SyncStatusPrefix
-
-	ctx, cancelFn := context.WithTimeout(context.Background(), flushTimeout)
-	defer cancelFn()
+	switch kind {
+	case logsv1.IngestBatch_ENTRY_KIND_ACCESS_LOG:
+		stream.Prefix = AccessSyncPrefix
+	case logsv1.IngestBatch_ENTRY_KIND_DECISION_LOG:
+		stream.Prefix = DecisionSyncPrefix
+	case logsv1.IngestBatch_ENTRY_KIND_UNSPECIFIED:
+		return errors.New("unspecified IngestBatch_EntryKind")
+	}
 
 	stream.Send = func(buf *z.Buffer) error {
 		kvList, err := badgerv4.BufferToKVList(buf)
 		if err != nil {
 			return err
 		}
 
-		keys := make([][]byte, len(kvList.Kv))
-		for i, kv := range kvList.Kv {
-			keys[i] = kv.Key
+		keys := make([][]byte, 0, len(kvList.Kv))
+		for _, kv := range kvList.Kv {
+			keys = append(keys, kv.Key)
```

Comment:
It was faster to use indexing rather than `append` in a benchmark I did a few Go versions ago. Not sure whether that has changed now.

Reply:
Ah. I assumed that pre-allocating a large enough capacity (can't determine exact length so used the sum of the two) would mean that `appends` would still just add to the end of the pre-existing array? To index, I'd need to set the length to (average-case) double the required number, and then the length would be all skewiff?

Reply:
Ohhh I see now. What I just described applied when I was still handling both prefixes in a single scan, I forgot to change in the update. Thanks for raising